### DISCUSSION TITLE
Second run of prisma migrate deploy fails

### DISCUSSION THREAD

**Discussion by jacob-regan-tempus** [prisma/prisma#12452] [Q&A]
I have my startup script for CI setup to run `prisma migrate deploy` before running my main app process. On first run, the migration was successful. On second run (with no schema changes to the database) I ran into the following error: `Error querying the database: db error: ERROR: relation "_prisma_migrations" already exists`. Wondering why this table collision happens since the table should persist to track schema changes over time.

**Comment by janpio** [comment]
For some reason it thinks it does not exist and needs to be created again. Can you share some more information about your setup? Prisma version, what kind of migration is in the first run and so on? What database and version? Do you have some special permissions/grants/user setup on that database?

**Reply by jacob-regan-tempus** [reply]
Hey, thanks for the response. I'm running prisma `2.25.0` with Postgres `10.18`. I had 5 local migrations from development in the `migrations` directory when I attempted running it in CI which involved basic table creation and then a couple follow up migrations to alter the tables. The first migration run with `migrate deploy` worked fine, and then the second threw that error about the existing `_prisma_migrations` relation. The machine user connected to the db should have proper permissions to execute all needed operations.

**Reply by jacob-regan-tempus** [reply]
Upgraded to `3.11.0` and still encountering the same issue. I traced the source code in this repo down as far as I could to the Migration engine kicking off an `applyMigration` and constructing a JSON RPC payload to a child process that would eventually execute a SQL query, but would you happen to know if that query ends up being `CREATE TABLE IF NOT EXISTS` for postgres or does it just run `CREATE TABLE` every time? Also curious if the syntax just isn't working for the version of Postgres I have running

**Reply by jacob-regan-tempus** [reply]
The reason I'd like to confirm that is that I can reproduce the error in my local postgres by running `CREATE TABLE _prisma_migrations`, but when I use the `IF NOT EXISTS` clause, it just produces a log that it is skipping that operation and continues.

**Reply by janpio** [reply]
The `_prisma_migrations` table is handled outside of the migrations files and represent the Migrations system status itself, so it is not reflected in the migrations itself. As such we also do not use `IF NOT EXISTS` there, as our code should only try to create it if it does not exist: https://github.com/prisma/prisma-engines/blob/4432e347e3d732abc520eccbd8653361b98178ea/migration-engine/connectors/sql-migration-connector/src/flavour/postgres.rs#L290-L305

So for some reason on the second `migrate deploy` it does not understand that the table got created the first time so it could check which migrations need to be run, and starts as if this was the first `migrate deploy` on this database - by creating an empty migration table, which fails.

**Reply by jacob-regan-tempus** [reply]
That makes sense. I was kind've guessing this was happening. I guess my question is: how is this workflow different from any other in performing a migration? If the client always tries to create this table on every migration, wouldn't this error always be thrown? Or does the client typically tear down the table after a successful migration run?

**Reply by janpio** [reply]
The Migration Engine usually can see that the table is already present and uses it to learn about the state of migrations in that database. Somehow this is not working for your database, and it thinks it is in a new environment where it has to start from nothing. I am trying to figure out internally what check could be going wrong here.

**Reply by jacob-regan-tempus** [reply]
Awesome, thank you. Anything you come up with would be super helpful

**Reply by janpio** [reply]
We had a look at that code, and it basically reads like this:
1. If `_prisma_migrations` exists: Exit OK
2. If there are non system tables: Exit with error message
3. Otherwise: `CREATE TABLE _prisma_migration`

(If you feel like reading some Rust code: https://github.com/prisma/prisma-engines/blob/f2e63a76e8eed99d99f0eb65596c7677b70a1e77/migration-engine/connectors/sql-migration-connector/src/sql_migration_persistence.rs#L16-L40)

And your second `migrate deploy` seems to fall straigt into 3) as it does not find the table but also no others (which is weird - you said it successfully created the tables, right?).

The only thing I can think of is that the database user you are runnig with can create the tables, but not see them later. Can you run `npx prisma db pull` with it and see what is output or happens to your schema file? That might be interesting if that also detects the database as empty.

**Reply by jacob-regan-tempus** [reply]
Thanks for linking the code. And yes, seems like its falling into that 3rd block. It is able to successfully create the tables on first run. I was able to log in to the db and confirm the schema was created. That would really surprise me if the user was not able to list the tables. Is it possible that `self.flavour().migrations_table_name()` evaluates to something unexpected? Seems unlikely...and you probably would have run into this before if it were the case, but its possible. I did notice that the `flavour` linked to was SQL. Is it possible that there is just a slight variation between that and Postgres in this case? My migration table is written with a leading underscore: `_prisma_migrations`. Could be that `self.flavour().migrations_table_name()` evaluates to `prisma_migrations` or something and so it doesn't catch it?

**Reply by janpio** [reply]
No, that is just an internal function name that then gets turned into the correct string. If this was a general problem with PostgreSQL, we doul have _many_ more people contacting us about this - which is also why I suspect some more weird root cause.

This might be worth trying as we could learn something:
The only thing I can think of is that the database user you are runnig with can create the tables, but not see them later. Can you run npx prisma db pull with it and see what is output or happens to your schema file? That might be interesting if that also detects the database as empty.

**Reply by jacob-regan-tempus** [reply]
Ok, we have an update here! Finally got it working. Ended up adding `&schema=<servicename>` in our postgres `DATABASE_URL` variable. It defaults to public when you don't provide it. For some reason this worked fine locally but did not work properly in our lower deployed environments.

**Reply by jacob-regan-tempus** [reply]
Only took us 2 weeks to get here 😅  thanks again for jumping in and assisting. The Rust code you pointed us to ended up giving us enough context to debug in the right direction.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
